simple_salesforce/api.py

```diff
@@ -383,34 +384,21 @@ def query_all(self, query, **kwargs):
         * query -- the SOQL query to send to Salesforce, e.g.
                    `SELECT Id FROM Lead WHERE Email = "waldo@somewhere.com"`
         """
-        def get_all_results(previous_result, **kwargs):
-            """Inner function for recursing until there are no more results.
 
-            Returns the full set of results that will be the return value for
-            `query_all(...)`
-
-            Arguments:
-
-            * previous_result -- the modified result of previous calls to
-                                 Salesforce for this query
-            """
-            if previous_result['done']:
-                return previous_result
-            else:
-                result = self.query_more(previous_result['nextRecordsUrl'],
-                                         identifier_is_url=True, **kwargs)
-                result['totalSize'] += previous_result['totalSize']
-                # Include the new list of records with the previous list
-                previous_result['records'].extend(result['records'])
-                result['records'] = previous_result['records']
-                # Continue the recursion
-                return get_all_results(result, **kwargs)
-
-        # Make the initial query to Salesforce
         result = self.query(query, **kwargs)
-        # The number of results might have exceeded the Salesforce batch limit
-        # so check whether there are more results and retrieve them if so.
-        return get_all_results(result, **kwargs)
+        all_records = []
+
+        while True:
+            all_records.extend(result['records'])
+            # fetch next batch if we're not done else break out of loop
+            if not result['done']:
+                result = self.query_more(result['nextRecordsUrl'],
+                                         True)
+            else:
+                break
+
+        result['records'] = all_records
```

Comment:
This is an attempt at backward compatibility, since this is basically what the code was doing. 

Not sure why the `done` or `totalSize` fields need to be returned, since `done` will always be `True` and `totalSize` should always be equal to `len(result['records'])` here. 

Note that unlike the original implementation and per [this PR](https://github.com/simple-salesforce/simple-salesforce/pull/133) I am simply returning the `totalSize`, which seems to be the right thing to do.